File: datastruct/avl/AVL.py
class Node(object):

    def __init__(self, key=None, value=None, left=None, right=None):
        self.key = key
        self.value = value
        self.left = left
        self.right = right
        self.height = 1
    
    def __str__(self):
        return str(self.key) + ': ' + str(self.value)

class AVL(object):

    def __init__(self):
        self.__size = 0
        self.__root = None
    
    def is_empty(self):
        '''AVL树是否为空'''
        return self.__size == 0
    
    def get_size(self):
        '''AVL树的大小'''
        return self.__size

    def get_height(self, node: Node)-> int:
        '''获取结点高度'''
        if node is None:
            return 0
        return node.height

    def __get_balanced_factor(self, node: Node)-> int:
        '''获取结点平衡因子'''
        if node is None:
            return 0
        return self.get_height(node.left) - self.get_height(node.right)

    def __right_rotate(self, node: Node) -> Node:
        '''右旋转操作'''
        left = node.left
        left_max = left.right

        left.right = node
        node.left = left_max

        node.height = 1 + max(self.get_height(node.left), self.get_height(node.right))
        left.height = 1 + max(self.get_height(left.left), self.get_height(left.right))

        return left

    def __left_rotate(self, node: Node)-> Node:
        '''左旋转操作'''
        right = node.right
        right_min = right.left

        right.left = node
        node.right = right_min

        node.height = 1 + max(self.get_height(node.left), self.get_height(node.right))
        right.height = 1 + max(self.get_height(right.left), self.get_height(right.right))

        return right

    def add(self, key, value):
        self.__root = self.__add(self.__root, key, value)

    def __add(self, node: Node, key, value):
        '''向以node为根的AVL树中添加（key，value）键值对'''
        if node is None:
            self.__size += 1
            return Node(key, value)
        
        if key < node.key:
            node.left = self.__add(node.left, key, value)
        else:
            node.right = self.__add(node.right, key, value)

        # 更新高度
        node.height = 1 + max(self.get_height(node.left), self.get_height(node.right))

        balance_factor = self.__get_balance_factor(node)

        if balance_factor > 1 and self.__get_balance_factor(node.left) >= 0:
            return self.right_rotate(node)
        if balance_factor < -1 and self.__get_balance_factor(node.right) <= 0:
            return self.left_rotate(node)
        if balance_factor > 1 and self.__get_balance_factor(node.left) < 0:
            node.left = self.left_rotate(node.left)
            return self.right_rotate(node)
        if balance_factor < -1 and self.__get_balance_factor(node.right) > 0:
            node.right = self.right_rotate(node.right)
            return self.left_rotate(node)   

        return node

    def get_node(self, key):
        '''获取键值为 key 的结点'''
        if key is None:
            return None
        return self.__get_node(self.__root, key)
    
    def __get_node(self, node: Node, key):
        '''从以 node 为根的AVL树中寻找键值为 key 的结点并返回'''
        if node is None:
            return None

        if key < node.key:
            return self.__get_node(node.left, key)
        elif key > node.key:
            return self.__get_node(node.right, key)
        else:
            return node

    def contains(self, key):
        node = self.get_node(key)

        if node is None:
            return False
        else:
            return True
    
    def get(self, key):
        node = self.get_node(key)

        if node is None:
            return None
        else:
            return node.value

    def set(self, key, value):
        node = self.get_node(key)

        if node is None:
            raise IndexError('%s does not exist!' % key)

        node.value = value
    
    def __min(self, node: Node):

        if node is None:
            raise IndexError('Tree is empty!')

        if node.left is not None:
            return self.__min(node.left)
        
        return node

    def remove(self, key):
        ret = self.get_node(key)
        if ret is not None:
            ret = ret.value
            self.__root = self.__remove(self.__root, key)
        return ret

    def __remove(self, node: Node, key):
        '''从以node为根的AVL树中删除键值为 key 的结点'''
        if node is None:
            raise IndexError('Tree is empty!')

        ret_node = None
        if key < node.key:
            node.left = self.__remove(node.left, key)
            ret_node = node
        elif key > node.key:
            node.right = self.__remove(node.right, key)
            ret_node = node
        else: # key == node.key
            if node.left is None:
                self.__size -= 1
                ret = node.right
                node.right = None
                ret_node = ret
            elif node.right is None:
                self.__size -= 1
                ret = node.left
                node.left = None
                ret_node = ret
            else:
                ret = self.__min(node.right)
                ret.right = self.__remove(node.right, ret.key)
                ret.left = node.left
                node.right = node.left = None
                ret_node = ret

        if ret_node is None:
            return None

        ret_node.height = 1 + max(self.get_height(ret_node.left), self.get_height(ret_node.right))

        balance_factor = self.__get_balance_factor(ret_node)

        if balance_factor > 1 and self.__get_balance_factor(ret_node.left) >= 0:
            ret_node = self.right_rotate(ret_node)
        if balance_factor < -1 and self.__get_balance_factor(ret_node.right) <= 0:
            ret_node = self.left_rotate(ret_node)
        if balance_factor > 1 and self.__get_balance_factor(ret_node.left) < 0:
            ret_node.left = self.left_rotate(ret_node.left)
            ret_node = self.right_rotate(ret_node)
        if balance_factor < -1 and self.__get_balance_factor(ret_node.right) > 0:
            ret_node.right = self.right_rotate(ret_node.right)
            ret_node = self.left_rotate(ret_node)           

        return ret_node

File: datastruct/set/AVL.java
import	java.util.ArrayList;

public class AVL<K extends Comparable <K>, V>{

    private class Node{
        public K key;
        public V val;
        public Node left;
        public Node right;
        public int height;

        public Node(K key, V val) {
            this.key = key;
            this.val = val;
            this.left = null;
            this.right = null;
            this.height = 1;
        }
    }

    private Node root;
    private int size;

    public AVL(){
        root = null;
        size = 0;
    }


    public boolean isEmpty() {
        return size == 0;
    }


    public int getSize() {
        return size;
    }

    // 向 AVL 树中添加元素（key，value）
    public void add(K key, V value) {
        root = add(root, key, value);
    }

    // 获取结点高度
    private int getHeight(Node node){
        if(node == null)
            return 0;

        return node.height;
    }

    // 计算平衡因子
    private int getBalanceFactor(Node node){
        if(node == null)
            return 0;

        return getHeight(node.left) - getHeight(node.right);
    }

    // 右旋转操作
    private Node rightRotate(Node node){
        Node left = node.left;
        Node leftMax = left.right;

        left.right = node;
        node.left = leftMax;

        node.height = 1 + Math.max(getHeight(node.left), getHeight(node.right));
        left.height = 1 + Math.max(getHeight(left.left), getHeight(left.right));

        return left;
    }

    // 左旋转操作
    private Node leftRotate(Node node){
        Node right = node.right;
        Node rightMin = right.left;

        right.left = node;
        node.right = rightMin;

        node.height = 1 + Math.max(getHeight(node.left), getHeight(node.right));
        right.height = 1 + Math.max(getHeight(right.left), getHeight(right.right));

        return right;
    }

    // 向以 node 为根的 AVL 树添加元素（key，value），递归实现
    // 返回添加元素后的根结点
    private Node add(Node node, K key, V value) {

        if(node == null){
            size++;
            return new Node(key, value);
        }

        if(key.compareTo(node.key) > 0 )
            node.right = add(node.right, key, value);
        else
            node.left = add(node.left, key, value);

        // 更新高度
        node.height = 1 + Math.max(getHeight(node.left), getHeight(node.right));

        // 计算平衡因子
        int balanceFactor = getBalanceFactor(node);


        // LL，形如 /，右旋转。左子树最大值转到右子树最小值
        if(balanceFactor > 1 && getBalanceFactor(node.left) >= 0)
            return rightRotate(node);

        // RR，形如 \，左旋转。右子树最小值转到左子树最大值
        if(balanceFactor < -1 && getBalanceFactor(node.right) <= 0)
            return leftRotate(node);

        // LR，形如 <，先对左子树做左旋转，再对结点做右旋转
        if(balanceFactor >  1 && getBalanceFactor(node.left) < 0){
            node.left = leftRotate(node.left);
            return rightRotate(node);
        }

        // RL,形如 >, 先对右子树右旋转，再对结点左旋转
        if(balanceFactor < -1 && getBalanceFactor(node.right) > 0){
            node.right = rightRotate(node.right);
            return leftRotate(node);
        }

        return node;
    }

    // 返回以 node 为根结点的二分搜索树中关键字为 key 的结点
    private Node getNode(Node node, K key){
        if(node == null)
            return null;

        if(key.compareTo(node.key) < 0)
            return getNode(node.left, key);
        else if(key.compareTo(node.key) > 0)
            return getNode(node.right, key);
        else // key.compareTo(node.key) == 0
            return node;
    }

    public boolean contain(K key) {
        return getNode(root, key) != null;
    }


    public V get(K key) {
        Node node = getNode(root, key);
        return node == null ? null : node.val;
    }

    public void set(K key, V value) {
        Node node = getNode(root, key);

        if(node == null)
            throw new IllegalArgumentException(key + " doesn't exist!");

        node.val = value;
    }

    // 返回以 node 为根结点的二分搜索树中的最小节点
    private Node min(Node node){
        if(node == null)
            throw new IllegalArgumentException("Tree is empty!");

        while(node.left != null)
            node = node.left;

        return node;
    }

    public V remove(K key) {
        Node node = getNode(root, key);
        if(node == null)
            return null;
        root = remove(root, key);

        return node.val;
    }

    // 从以 node 为根结点的二分搜索树中删除 key 所在的结点
    // 返回删除结点后的根结点
    private Node remove(Node node, K key) {

        Node retNode;
        if(key.compareTo(node.key) < 0){
            node.left = remove(node.left, key);
            retNode = node;
        }
        else if(key.compareTo(node.key) > 0){
            node.right = remove(node.right, key);
            retNode = node;
        }
        else{ // key.compareTo(node.key) == 0
            if(node.left == null){
                Node ret = node.right;
                node.right = null;
                size--;
                retNode = ret;
            }
            else if(node.right == null){
                Node ret = node.left;
                node.left = null;
                size--;
                retNode = ret;
            }
            else{
                Node ret = min(node.right);
                ret.right = remove(ret, key);
                ret.left = node.left;
                node.left = null;
                node.right = null;
                retNode = ret;
            }
        }

        if(retNode == null)
            return null;

        // 更新高度
        retNode.height = 1 + Math.max(getHeight(retNode.left), getHeight(retNode.right));

        // 计算平衡因子
        int balanceFactor = getBalanceFactor(retNode);


        // LL，形如 /，右旋转。左子树最大值转到右子树最小值
        if(balanceFactor > 1 && getBalanceFactor(retNode.left) >= 0)
            return rightRotate(retNode);

        // RR，形如 \，左旋转。右子树最小值转到左子树最大值
        if(balanceFactor < -1 && getBalanceFactor(retNode.right) <= 0)
            return leftRotate(retNode);

        // LR，形如 <，先对左子树做左旋转，再对结点做右旋转
        if(balanceFactor >  1 && getBalanceFactor(retNode.left) < 0){
            retNode.left = leftRotate(retNode.left);
            return rightRotate(retNode);
        }

        // RL,形如 >, 先对右子树右旋转，再对结点左旋转
        if(balanceFactor < -1 && getBalanceFactor(retNode.right) > 0){
            retNode.right = rightRotate(retNode.right);
            return leftRotate(retNode);
        }

        return retNode;
    }

    // 判断是否满足二分搜索树性质
    public boolean isBST(){
        ArrayList<K> keys = new ArrayList<> ();
        inOrder(root, keys);
        for(int i = 1; i < keys.size(); i++){
            if(keys.get(i - 1).compareTo(keys.get(i)) > 0)
                return false;
        }
        return true;
    }

    private void inOrder(Node node, ArrayList<K> keys){
        if(node == null)
            return;

        inOrder(node.left, keys);
        keys.add(node.key);
        inOrder(node.right, keys);
    }

    // 判断是否满足平衡二叉树
    public boolean isBalanced() {
        return isBalanced(root);
    }

    private boolean isBalanced(Node node){
        if(node == null)
            return true;

        int balanceFactor = getBalanceFactor(node);
        if(Math.abs(balanceFactor) > 1)
            return false;
        return isBalanced(node.left) && isBalanced(node.right);
    }
}
